pkg/scheduler/queue/tenant_queues_test.go

```diff
@@ -32,55 +71,126 @@ func TestQueues(t *testing.T) {
 	assert.Nil(t, tenant)
 	assert.NoError(t, err)
 
-	// Add queues: [one]
-	qOne := getOrAdd(t, qb, "one", 0)
-	lastTenantIndex = confirmOrderForQuerier(t, qb, "querier-1", lastTenantIndex, qOne, qOne)
+	// Add tenant queues: tenant "one"
+	err = qb.tenantQuerierAssignments.createOrUpdateTenant("one", 0)
+	assert.NoError(t, err)
+
+	tenantOne := qb.tenantQuerierAssignments.tenantsByID["one"]
+
+	// Queue enough objects for tenant "one"
+	err = qb.enqueueObjectsForTests(tenantOne.tenantID, 10)
+	assert.NoError(t, err)
+	assert.NoError(t, isConsistent(qb))
+
+	//queuePathOne := QueuePath{"root", "one"}
+	expectedDequeueVals := []dequeueVal{
+		{buildExpectedObject(tenantOne.tenantID, 0), tenantOne},
+		{buildExpectedObject(tenantOne.tenantID, 1), tenantOne},
+	}
+	lastTenantIndex = assertExpectedValuesOnDequeue(t, qb, lastTenantIndex, "querier-1", expectedDequeueVals)
+
+	// Add tenant two
+	err = qb.tenantQuerierAssignments.createOrUpdateTenant("two", 0)
+	assert.NoError(t, err)
+
+	tenantTwo := qb.tenantQuerierAssignments.tenantsByID["two"]
+
+	err = qb.enqueueObjectsForTests(tenantTwo.tenantID, 10)
+	assert.NoError(t, err)
+	assert.NoError(t, isConsistent(qb))
+
+	expectedDequeueVals = []dequeueVal{
+		{buildExpectedObject(tenantTwo.tenantID, 0), tenantTwo},
+		{buildExpectedObject(tenantOne.tenantID, 2), tenantOne},
+		{buildExpectedObject(tenantTwo.tenantID, 1), tenantTwo},
+		{buildExpectedObject(tenantOne.tenantID, 3), tenantOne},
+	}
+
+	lastTenantIndex = assertExpectedValuesOnDequeue(t, qb, lastTenantIndex, "querier-1", expectedDequeueVals)
+
+	expectedDequeueVals = []dequeueVal{
+		{buildExpectedObject(tenantOne.tenantID, 4), tenantOne},
+		{buildExpectedObject(tenantTwo.tenantID, 2), tenantTwo},
+		{buildExpectedObject(tenantOne.tenantID, 5), tenantOne},
+	}
+	_ = assertExpectedValuesOnDequeue(t, qb, -1, "querier-2", expectedDequeueVals)
+
+	//[one two three]
+	// confirm fifo by adding a third tenant queue and iterating to it
+	err = qb.tenantQuerierAssignments.createOrUpdateTenant("three", 0)
+	assert.NoError(t, err)
 
-	// [one two]
-	qTwo := getOrAdd(t, qb, "two", 0)
+	tenantThree := qb.tenantQuerierAssignments.tenantsByID["three"]
 
-	lastTenantIndex = confirmOrderForQuerier(t, qb, "querier-1", lastTenantIndex, qTwo, qOne, qTwo, qOne)
-	confirmOrderForQuerier(t, qb, "querier-2", -1, qOne, qTwo, qOne)
+	err = qb.enqueueObjectsForTests(tenantThree.tenantID, 10)
+	assert.NoError(t, err)
+	assert.NoError(t, isConsistent(qb))
 
-	// [one two three]
-	// confirm fifo by adding a third queue and iterating to it
-	qThree := getOrAdd(t, qb, "three", 0)
+	expectedDequeueVals = []dequeueVal{
```

Comment:
quick comment on _why_ we expect these to bet the next three values - our querier 1 last tenant index is for tenant 2, and tenant three got added to the back end of the rotation after two and before we return back to one

Reply:
Small (but important imo) correction -- at this point, `lastTenantIndex` for `querier-1` points to tenant _one_, since incrementing happens _during_ the dequeue operation. I got got a few times by off-by-one errors because I forgot where I was incrementing 😅